Docker setup

BlackVegetable opened the pull request:
Our fork off of twemproxy is simply to allow
us to build our own Dockerhub image.  There is no
authoritative twemproxy image on Dockerhub at
this time.

BlackVegetable committed:
First pass at setting up Docker here.

Our fork off of twemproxy is simply to allow
us to use our own configuration files within
our own image in Docker.

BlackVegetable committed:
typo

BlackVegetable committed:
+curl dependency

Turns out Ubuntu doesn't come with curl installed
by default.  We need it for stuff.

BlackVegetable committed:
Follow http links

Our curl command needs to follow redirects to download the
github release.

BlackVegetable committed:
Use local upstream

It was silly to be pulling from an official release when
we were on the fork to begin with!

BlackVegetable committed:
Added configuration files

These configuration files are all development
environment configurations.  They need to be
modified for use in staging or production.
But, as is, the circle build should complete
and should allow for local testing.

BlackVegetable committed:
Updated dependencies

I had kept curl in there when we didn't need it.
Also, I was missing apt-utils which seems to be
needed to avoid a warning/error message and probably
build errors.

BlackVegetable committed:
Trying to install apt-utils

Perhaps installing apt-utils has to be the
very first thing attempted?

BlackVegetable committed:
add necessary args

BlackVegetable committed:
update first

BlackVegetable committed:
cd to repo root

BlackVegetable committed:
try using WORKDIR

BlackVegetable committed:
Copy entire repo.

Rather than worrying about copying indvidiual files, I can copy
the entire repository at once. Durrr!

BlackVegetable committed:
rm ignore file for make success

BlackVegetable committed:
yaml -> yml

BlackVegetable commented:
This will have tests before it is merged.

BlackVegetable committed:
Moving to deploybot

We will use deploybot for the distributed
caching project after all.  This means we
will have a different pattern for passing
variables and files around.

Also, we don't want to write our own tests
for something we aren't modifying and that
has passing tests upstream.  So, we are
going to have a dummy test that should
always pass to make circle happy and not
notify us in Slack that we have problems
due to how our Slack integrations work with
circle.

BlackVegetable committed:
better fake test

BlackVegetable committed:
rm unused conf files

We are going to need to generate the configuration
files for this within our Deploybot. This is the case
because the list of servers and port numbers will need
to be dynamically generated. Thus, there is no need
to keep any of them in this repo.

terrhorn commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# Start Twemproxy
+# Use default configuration file location/name.
+# Use mbufsize from ansible commandline variable.
+ENTRYPOINT src/nutcracker
```
Nit: we usually put these in the exec form format (which is preferred) like so: `ENTRYPOINT ["src/nutcracker"]`

BlackVegetable replied:
Ok.

terrhorn commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+FROM ubuntu:16.04
```
Did we decide that we wanted to pull from the base Ubuntu image here rather than a pre-existing Keen image?

BlackVegetable replied:
I don't know which Keen image would be appropriate.  That would likely be wiser.  Do you have a suggestion?

terrhorn replied:
Let's just leave it as-is for now since the JDK isn't required. We can revisit in the future if necessary.

gorzell replied:
I think using the stock ubuntu base for this is fine.

terrhorn commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+  override:
+    - echo 0
+
+database:
```
Probably worth adding a comment above `database:` that we generally include here:

```
# Use this section to setup tests external dependencies, not just databases.
```

BlackVegetable replied:
Will do.

terrhorn commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+
+test:
+  override:
+    - echo 0
```
Consider adding a comment here to explain why we're echoing `0` rather than running real tests.

BlackVegetable replied:
Right. I should do that.

gorzell commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+# Setup the dependencies
+RUN apt-get update -y
+RUN apt-get install -y apt-utils
```
These should all be one run command.  That way you don't generate unnecessary layers. https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#/sort-multi-line-arguments

```
RUN apt-get update -y && \
 apt-get install -y apt-utils libtool make automake
```

BlackVegetable replied:
When I did this, the latter tools spewed warnings because `apt-utils` is used to properly configure the later tools.  When I split it into multiple commands, that warning goes away.

gorzell replied:
Still would follow the recommendations they make and work to minimize the number of RUN commands.

BlackVegetable replied:
Ok, fair enough.

gorzell commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+  services:
+    - docker
+
+test:
```
`test` goes after `database`.  You want to try and keep these in the order they happen in circle. https://circleci.com/docs/manually/#overview

gorzell replied:
Why can't you do a real test here?

BlackVegetable replied:
We could recreate the upstream tests but I'm not sure what the point of that is.  We aren't adding code changes to this repo.

gorzell replied:
How about testing that what you installed actually starts and is available on the ports you specified?

BlackVegetable replied:
I didn't know that could be tested at image-build time.  Originally that was the main test I was considering.  I'll need to ask @terryhorner how that would be accomplished.

terrhorn replied:
We discussed this during our DevOps sync. Testing containers in Circle needs to be a part of a ticket/project. For now, a comment that describes why we're doing what we're doing will suffice.

terrhorn replied:
We found a simple example of testing via `curl` here: https://circleci.com/docs/docker/#deployment-to-a-docker-registry

gorzell commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+  package:
+    branch: /.*/
+    commands:
+      - docker build -t keen/twemproxy:$CIRCLE_BUILD_NUM .
```
Build the image in the `dependencies` section so that it can be started in the `database` section and then tested in the `test` section.

gorzell replied:
https://circleci.com/docs/configuration/

gorzell commented on circle.yml:
```diff
@@ -0,0 +1,19 @@
+    commands:
+      - docker build -t keen/twemproxy:$CIRCLE_BUILD_NUM .
+      - docker tag -f keen/twemproxy:$CIRCLE_BUILD_NUM keen/twemproxy:$CIRCLE_BRANCH
+      - docker push keen/twemproxy
```
the `docker push` is the only thing that should remain in this section.

terrhorn replied:
And `docker login`.

gorzell commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    make
+
+# Expose Twemproxy Ports
+EXPOSE 22122 22222
```
Use the array format here too.

BlackVegetable replied:
Ah, I didn't even know that was possible.  Thanks! I will do that.

gorzell replied:
https://docs.docker.com/engine/reference/builder/#expose

BlackVegetable replied:
Turns out those docs are stating you can use multiple ports in one line, optionally. Not: you can use array notation to express ports to expose.  I've tried to get the latter to work several times and have looked online, and failed to discover anyone using it.  I'm pretty sure it isn't used for this.

gorzell commented:
I am unclear what part of this uses "our own config files".

BlackVegetable commented:
You are right. There is nothing here that has anything to do with our config files.  I'm running on little sleep and am getting sloppy.  I'll update the PR comment to remove that old reference.  The only thing this is doing for us now is giving us a base twemproxy image to use because there isn't one in Dockerhub that is authoritative.

BlackVegetable committed:
refactor w.r.t code review

Changed some formatting to match preferred
standards from a code review.

BlackVegetable committed:
Refactor circle.yml

Clean this up to have better conventions
and comments explaining our "test".

BlackVegetable committed:
Now with real testing

We found this example at:
https://circleci.com/docs/docker/

BlackVegetable committed:
fix circle syntax

BlackVegetable committed:
docker syntax fix

BlackVegetable committed:
Fixed EXPOSE

EXPOSE does not seem to support
the array syntax.

BlackVegetable committed:
+build number

So if you don't refer to the build number
in the tests, it will attempt to use "latest"
which we don't have (and don't want to use anyway
for new builds.) So even in the tests you need
to specify which *exact* image to use, which in this
case means using the $CIRCLE_BUILD_NUMBER variable.
I should've thought of that. :)

BlackVegetable committed:
test build

This is a test build to see what that HTTP response
is in circle. This is working for me locally but is
returning a cURL error code fo 56 in Circle.

Hopefully this will give us some insight into what is
going on.

BlackVegetable committed:
Disabled Test

We need to debug this in staging as well as have a
working build for other deployment components to
use to test them against.

BlackVegetable committed:
typo

BlackVegetable committed:
typo2

BlackVegetable commented:
@terryhorner Given that we need this for testing deploybot in staging, and the test will be easier to debug in staging (or might not show up there, which would be valuable information as well), could you give this a look again despite the disabled test?

terrhorn reviewed (approved):
This is approved for testing in staging only. We need to find a solution to the `echo 0` prior to moving to production.

BlackVegetable commented:
Testing in staging is the plan. Thanks.
